File: 040.CombinationSumII/MarkCSIE/CombinationSumII/src/CombinationSumII.java

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class CombinationSumII {

    public static void main(String[] args) {
        CombinationSumII test = new CombinationSumII();
        int[] testCase1 = new int[]{10, 1, 2, 7, 6, 1, 5};
        List<List<Integer>> testAnswer1 = test.combinationSum2(testCase1, 8);
        System.out.println(testAnswer1.toString());
    }

    public List<List<Integer>> recursiveCombinationSum2(int[] candidates, int target) {
        List<List<Integer>> combinationList = new ArrayList<>();
        for (int i = 0; i < candidates.length; i++) {
            int number = candidates[i];
            if (i > 0 && number == candidates[i - 1]) {
                continue;
            }
            if (target - number > 0) {
                int[] subCandidates = Arrays.copyOfRange(candidates, i + 1, candidates.length);
                List<List<Integer>> subCombinationList = recursiveCombinationSum2(subCandidates, target - number);
                for (List<Integer> combination : subCombinationList) {
                    combination.add(0, number);
                }
                combinationList.addAll(subCombinationList);
            } else if (target - number == 0) {
                List<Integer> combination = new ArrayList<>();
                combination.add(number);
                combinationList.add(combination);
                break;
            } else {
                break;
            }
        }
        return combinationList;
    }

    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        return recursiveCombinationSum2(candidates, target);
    }
}

File: 040.CombinationSumII/wyrj/CombinationSumII.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum2 = function(candidates, target) {
    var i, j, k, back, temp;
    var table = [];
    candidates.sort(function(lhs, rhs) {
        return lhs - rhs;
    });
    for (i = 1; i <= target; i++) {
        table[i] = [];
        for (j = 0; j < candidates.length; j++) {
            table[i][j] = [];
            n = candidates[j];
            if (0 < j) {
                if (candidates[j - 1] !== n) {
                    table[i][j] = table[i][j - 1].slice(0);   
                } else {
                    for (k = 0; k < table[i][j - 1].length; k++) {
                        temp = table[i][j - 1][k];
                        if (temp[temp.length - 1] !== n) {
                            table[i][j].push(temp);
                        }
                    }
                }
            }
            if (n > i) {
                continue;
            } else if (n === i) {
                table[i][j].push([n]);
            } else if (0 !== j) {
                back = table[i - n][j - 1];
                for (k = 0; k < back.length; k++) {
                    temp = back[k].slice(0);
                    temp.push(n);
                    table[i][j].push(temp);
                }
            }
        }
    }
    return table[target][candidates.length - 1];
};